File: src/main/java/by/intexsoft/forum/entity/TopicDiscussRequest.java
package by.intexsoft.forum.entity;

import javax.persistence.*;
import java.sql.Timestamp;

import static javax.persistence.FetchType.EAGER;
import static javax.persistence.FetchType.LAZY;

/**
 * Describes 'topic_discuss_requests' table in DB
 */
@Entity
@Table(name = "topic_discuss_requests")
public class TopicDiscussRequest extends AbstractVersionEntity {
    private static final long serialVersionUID = 2114056750097120018L;

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "requested_by", updatable = false, nullable = false)
    public User requestedBy;

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "approved_by", insertable = false)
    public User approvedBy;

    @Column(name = "approved_at", insertable = false)
    public Timestamp approvedAt;

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "topic_id", updatable = false)
    public Topic inTopic;

    @ManyToOne(fetch = EAGER)
    @JoinColumn(name = "status_id", nullable = false)
    public Status status;
}

File: client/src/app/shared/entity/topic-discuss-request.ts
import { Topic } from './topic';
import { User } from './user';
import { AbstractVersioningEntity } from './abstract/abstract-versioning-entity';
import { Status } from './status';

export class TopicDiscussRequest extends AbstractVersioningEntity {
  requestedBy: User;
  approvedBy: User;
  approvedAt: Date;
  requestedById: number;
  inTopic: Topic;
  status: Status;
}

const PENDING = 'PENDING';
const APPROVED = 'APPROVED';
const REJECTED = 'REJECTED';

export class StatusConst {
  static get PENDING() {
    return PENDING;
  }

  static get APPROVED() {
    return APPROVED;
  }

  static get REJECTED() {
    return REJECTED;
  }
}
